File: ella/week7/1780.py
# 백로그
import sys
input = sys.stdin.readline
n = int(input())

p = [list(map(int, input().split())) for _ in range(n)]

answer = [0,0,0]

def det(x, y, z):
    global answer
    num = p[x][y]
    for i in range(x, x+z):
        for j in range(y, y+z):
            if num != p[i][j]:
                for a in range(3):
                    for b in range(3):
                        det(x + (z//3*a), y + (z//3*b), z//3)
                return # 종이에 채워진 숫자가 달라 9개로 자르고 종료

    # 종이에 채워진 숫자가 모두 같은 경우    
    if num == -1:
        answer[0] += 1
    elif num == 0:
        answer[1] += 1
    else:
        answer[2] += 1

det(0, 0, n)
print(answer[0])
print(answer[1])
print(answer[2])

File: leo/week7/1780.js
// Run by Node.js
const readline = require("readline");

(async () => {
  let rl = readline.createInterface({ input: process.stdin });

  let input = [];
  for await (const line of rl) {
    input.push(line);
  }

  let N = parseInt(input[0]);
  let paper = [];
  for (let i = 1; i <= N; i++) {
    paper.push(input[i].trim().split(" ").map(Number));
  }

  let counts = { "-1": 0, 0: 0, 1: 0 };

  function checkSameNumber(x, y, size) {
    let firstNumber = paper[x][y];
    for (let i = x; i < x + size; i++) {
      for (let j = y; j < y + size; j++) {
        if (paper[i][j] !== firstNumber) {
          return false;
        }
      }
    }
    return true;
  }

  function divideAndConquer(x, y, size) {
    if (checkSameNumber(x, y, size)) {
      counts[paper[x][y]]++;
    } else {
      let newSize = size / 3;
      for (let i = 0; i < 3; i++) {
        for (let j = 0; j < 3; j++) {
          divideAndConquer(x + i * newSize, y + j * newSize, newSize);
        }
      }
    }
  }

  divideAndConquer(0, 0, N);

  console.log(counts["-1"]);
  console.log(counts["0"]);
  console.log(counts["1"]);

  process.exit();
})();
